hack/test: basic support for byte based updates

bfredl opened the pull request:
test if we can use `inserted_bytes` from vim 8:s `change.c` (ref #10171) to make common update events more detailed.  Test script (works best with https://github.com/bfredl/nvim-luadev ):
```
local print = require('luadev').print

local function cb(nam, ...)
  local text = vim.inspect({...})
  vim.schedule(function () print(nam, text) end)
end

vim.api.nvim_buf_attach(0, false, {on_lines=cb, on_bytes=cb})
```

Seems to work with most small insert mode and normal mode changes (expect small put:s).

bfredl committed:
lua driven syntax higlighting (VERY ROUGH prototype)

bfredl committed:
hack/test: basic support for byte based updates

bfredl commented:
By hacking the memline further I _think_ we can get out the byte count of the deleted/replaced text in `lines_cb` quite easily. This would greatly simplify tree editing in tree-sitter.

 In extension, at every time the byte length is counted the exact text is also available, which means we could also could calculate the UTF-16 size for LSP servers that need it without any extra buffering.

justinmk commented on src/nvim/api/buffer.c:
```diff
@@ -142,6 +142,13 @@ Boolean nvim_buf_attach(uint64_t channel_id,
       }
       cb.on_lines = v->data.luaref;
       v->data.integer = LUA_NOREF;
+    } else if (is_lua && strequal("on_bytes", k.data)) {
```
Would on_data be an incorrect name? We use that elsewhere.

bfredl replied:
This entire API is subject to change, this is just the lazy "what misc1.c does internally" hack. Probably, for consumers that opt-in to byte resolution, there should be a singe event with `pos_T start, pos_T old_end, pos_T new_end` values (+ total byte count of the deleted region, though actually we should do that first, I will open a separate PR).

bfredl committed:
query list test

bfredl committed:
tree-sitter: update vendored tree-sitter runtime

tree-sitter/tree-sitter commit a7824cf59d8b077f929ea47ca5bf107102fcd5ac

Included files are:
lib/include/tree-sitter/*.h
lib/src/*.[ch]
lib/src/unicode/*
LICENSE

bfredl committed:
query: use iterator

bfredl committed:
port over the basic prototype from the old sheet based solution

bfredl committed:
start proper luahl api

bfredl committed:
multiline test (like /* ... \n ... */ comment)

bfredl committed:
fix crash

bfredl committed:
test: more highlights!

bfredl committed:
query inspection

bfredl committed:
WIP: return full match

bfredl committed:
initial predicate support

bfredl committed:
handle (eq? at-aa at-bb)

bfredl committed:
use ts_query_cursor_remove_match

bfredl committed:
objectify highlight moddule

bfredl committed:
changes (does not work yet)

bfredl committed:
fix

bfredl committed:
sketch multiline update

for the situation

/* <- remove the "*"
 ..
 ..
*/

the entire range needs to be rehighlighted

bfredl committed:
unused sketch for redraw_buf_range_later, save for later

bfredl committed:
allow query to be changed

bfredl committed:
FileType "plugin" for c highlights (with query reload support)

bfredl committed:
fix: memory error on blank line + handle invalid botline

bfredl committed:
allow @UpperCase for direct highlight

bfredl committed:
use integer ids for capatures always.

also allow @HLGroup.specifier i e

((binary_expression left: (identifier) @WarningMsg.left right: (identifier) @WarningMsg.right) (eq? @WarningMsg.left @WarningMsg.right))

bfredl committed:
factor out the predicate logic from highlight code

bfredl committed:
missing include (upstream?)

bfredl committed:
roname

bfredl committed:
FIXUP

bfredl committed:
initial test

bfredl committed:
use it in tree-sitter

bfredl committed:
add nvim_get_hl_id_by_name, fix improper attr id based interface.

bfredl committed:
redraw correctly with split windows

bfredl committed:
gåå

bfredl committed:
FIXUP: should on_start always??

bfredl committed:
debug/perf: simple edit/redraw count

bfredl committed:
debug/perf: more advanced edit/redraw count
